Addon-docs: Fix exotic React components in Source block

andezzat opened the pull request:
## The Problem

React Memo & ForwardRef exotic components would show up in the code snippet of storybook as `<[object Object]>` because it didn't know where to lookup the inner render function's name.

This is a bug in the library Storybook uses to render JSX into a string: https://github.com/algolia/react-element-to-jsx-string/issues/352

## What I did

I stole the default `displayNameFn` from the original library [here](https://github.com/algolia/react-element-to-jsx-string/blob/4974512273c86c72415376fea89f9d6e07e2b2e5/src/parser/parseReactElement.js#L15) 
And added what it needed to fetch Memo & ForwardRef component's names.

## How to test

- Is this testable with Jest or Chromatic screenshots?
Maybe? I have no idea
- Does this need a new example in the kitchen sink apps?
No
- Does this need an update to the documentation?
No

If your answer is yes to any of these, please make sure to include it in your PR.

<!--

Everybody: Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `master` branch as part of the release process, so you shouldn't need to worry about this.

Maintainers: Please tag your pull request with at least one of the following:
`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

-->

andezzat committed:
fix(react/jsxDecorator): make rendered component's displayName work w/ memo & forwardRef exotic components

Prev. would just do as <[object Object]>.. because it didn't know where to lookup the inner render function's name.

andezzat committed:
fix(addon-docs): add any types to props so it builds

andezzat committed:
Delete .vscode settings (woops!)

ndelangen committed:
fix versions

ndelangen committed:
fix snapshot

andezzat commented on addons/docs/src/frameworks/react/jsxDecorator.test.tsx:
```diff
@@ -103,7 +103,7 @@ describe('renderJsx', () => {
           'item 19'
         ]}
-       />	
+       />
```
😂 
For some reason linter worked first time, but not when I forked 🤷 
Sorry!

shilman committed:
Add React.memo dynamic snippet to official-storybook

shilman reviewed (changes requested):
Hey @andezzat !! Thanks so much for putting this together -- will make a ton of people happy. Unfortunately, I added a test case to `official-storybook` and it doesn't seem to be working end-to-end. You can find that in the latest commit on this branch.

Would you mind:
- looking into this specific case
- adding more test cases to official-storybook
- reconciling the official-storybook behavior and the jest test cases (look great, but somehow don't line up)

andezzat committed:
fix(official-storybook): fix displayName for memo & add forward-ref example

Component should be fed directly to React.memo|forwardRef functions so that displayName is set. Previously was passing an anonymous arrow fn...

andezzat committed:
fix(official-storybook): use existing forward-ref story

shilman commented on examples/official-storybook/stories/addon-docs/react-forward-ref.stories.js:
```diff
@@ -0,0 +1,16 @@
+
+const ButtonWithForwardRef = React.forwardRef(DocgenButton);
+
+export default {
```
can we delete this file now?

andezzat replied:
Woopsie! 😅

andezzat committed:
fix(official-storybook): cleanup

shilman reviewed (approved):
LGTM! 💯 💯 💯

shilman committed:
Merge branch 'next' into fix/react-component-name

ghengeveld commented:
This appears to be broken:

<img width="502" alt="Screenshot 2020-10-09 at 14 05 32" src="https://user-images.githubusercontent.com/321738/95581002-8920b300-0a38-11eb-9d40-efff975c3111.png">

It seems to originate from `ForwardRefButtonInnerPropTypes` which wraps a button component directly in `forwardRef`, which is not allowed.

shilman commented:
cc @andezzat can you take a look ☝️ \

andezzat commented:
Ah yes, I should've made a new component w/ proper `forwardRef` fn args than passing `BaseButton` as-is 😂
PR: #12733
